src/obj_char_transfer.c

```diff
@@ -0,0 +1,872 @@
+    } else {
+        ObjCharTransferTask_G2dLoadImageMappingByScreen(task, NNS_G2D_VRAM_TYPE_2DMAIN);
+        ObjCharTransferTask_G2dLoadImageMappingByScreen(task, NNS_G2D_VRAM_TYPE_2DSUB);
+    }
+}
+
+static void ObjCharTransferTask_G2dLoadImageMappingByScreen(ObjCharTransferTask *task, NNS_G2D_VRAM_TYPE vram) {
+    int mappingType = task->shouldUpdateMappingTypeFromHW ? ObjCharTransferTask_SetMappingTypeFromHW(task, vram) : vram;
+    u32 baseAddr;
+    if (vram == NNS_G2D_VRAM_TYPE_2DMAIN) {
+        baseAddr = task->baseAddrMain;
+        GX_GetOBJVRamModeChar();
+    } else {
+        baseAddr = task->baseAddrSub;
+        GXS_GetOBJVRamModeChar();
+    }
+    if (mappingType == GX_OBJVRAMMODE_CHAR_2D) {
+        NNS_G2dLoadImage2DMapping(task->charData, baseAddr, vram, &task->imageProxy);
+    } else {
+        NNS_G2dLoadImage1DMapping(task->charData, baseAddr, vram, &task->imageProxy);
+    }
+    if (vram == NNS_G2D_VRAM_TYPE_2DMAIN) {
+        GX_GetOBJVRamModeChar();
+    } else {
+        GXS_GetOBJVRamModeChar();
+    }
+}
+
+static void ObjCharTransferTask_G2dLoadImageMappingVramTransfer(ObjCharTransferTask *task) {
+    NNS_G2dInitImageProxy(&task->imageProxy);
+    if (task->vram != NNS_G2D_VRAM_TYPE_BOTH) {
+        ObjCharTransferTask_G2dLoadImageMappingVramTransferByScreen(task, task->vram);
+    } else {
+        ObjCharTransferTask_G2dLoadImageMappingVramTransferByScreen(task, NNS_G2D_VRAM_TYPE_2DMAIN);
+        ObjCharTransferTask_G2dLoadImageMappingVramTransferByScreen(task, NNS_G2D_VRAM_TYPE_2DSUB);
+    }
+}
+
+static void ObjCharTransferTask_G2dLoadImageMappingVramTransferByScreen(ObjCharTransferTask *task, NNS_G2D_VRAM_TYPE vram) {
+    int mappingType = task->shouldUpdateMappingTypeFromHW ? ObjCharTransferTask_SetMappingTypeFromHW(task, vram) : vram;
+    #pragma unused(mappingType)
+    u32 baseAddr;
+    if (vram == NNS_G2D_VRAM_TYPE_2DMAIN) {
+        baseAddr = task->baseAddrMain;
+        GX_GetOBJVRamModeChar();
+    } else {
+        baseAddr = task->baseAddrSub;
+        GXS_GetOBJVRamModeChar();
+    }
+    NNS_G2dLoadImageVramTransfer(task->charData, baseAddr, vram, &task->imageProxy);
+    if (vram == NNS_G2D_VRAM_TYPE_2DMAIN) {
+        GX_GetOBJVRamModeChar();
+    } else {
+        GXS_GetOBJVRamModeChar();
+    }
+}
+
+static ObjCharTransferTask *ObjCharTransfer_GetFreeTask(void) {
+    for (int i = 0; i < sObjCharTransferTasksManager->max; ++i) {
+        if (sObjCharTransferTasksManager->tasks[i].state == OBJ_CHAR_TRANSFER_TASK_INIT) {
+            return &sObjCharTransferTasksManager->tasks[i];
+        }
+    }
+
+    return NULL;
+}
+
+static void ObjCharTransfer_GetVramCapacityByBank(void) {
+    switch (GX_GetBankForOBJ()) {
+    case GX_VRAM_OBJ_NONE:
+        sObjCharTransferTasksManager->vramCapacityMain = 0;
+        break;
+    case GX_VRAM_OBJ_16_F:
+    case GX_VRAM_OBJ_16_G:
+        sObjCharTransferTasksManager->vramCapacityMain = 16 * 1024;
```

Comment:
I'd suggest some constants here as it'd make it cleaner, but I don't actually know what to suggest